File: leader/src/main/scala/io/shinto/amaterasu/leader/execution/JobLoader.scala
package io.shinto.amaterasu.leader.execution

import java.util.concurrent.BlockingQueue

import io.shinto.amaterasu.common.dataobjects.ActionData
import io.shinto.amaterasu.common.logging.Logging
import io.shinto.amaterasu.enums.ActionStatus
import io.shinto.amaterasu.leader.dsl.{GitUtil, JobParser}
import org.apache.curator.framework.CuratorFramework
import org.apache.zookeeper.CreateMode

import scala.collection.JavaConverters._

/**
  * Created by roadan on 3/7/16.
  */
object JobLoader extends Logging {

  def loadJob(src: String, branch: String, jobId: String, client: CuratorFramework, attempts: Int, actionsQueue: BlockingQueue[ActionData]): JobManager = {

    // creating the jobs znode and storing the source repo and branch
    client.create().withMode(CreateMode.PERSISTENT).forPath(s"/$jobId")
    client.create().withMode(CreateMode.PERSISTENT).forPath(s"/$jobId/repo", src.getBytes)
    client.create().withMode(CreateMode.PERSISTENT).forPath(s"/$jobId/branch", branch.getBytes)

    val maki: String = loadMaki(src, branch)

    val jobManager: JobManager = createJobManager(maki, jobId, client, attempts, actionsQueue)

    //jobManager.start()
    jobManager

  }

  def createJobManager(maki: String, jobId: String, client: CuratorFramework, attempts: Int, actionsQueue: BlockingQueue[ActionData]): JobManager = {

    val jobManager = JobParser.parse(
      jobId,
      maki,
      actionsQueue,
      client,
      attempts
    )
    jobManager
  }

  def loadMaki(src: String, branch: String): String = {

    // cloning the git repo
    log.debug(s"getting repo: $src, for branch $branch")
    GitUtil.cloneRepo(src, branch)

    // parsing the maki.yaml and creating a JobManager to
    // coordinate the workflow based on the file
    val maki = JobParser.loadMakiFile()
    maki
  }

  def reloadJob(jobId: String, client: CuratorFramework, attempts: Int, actionsQueue: BlockingQueue[ActionData]) = {

    //val jobState = client.getChildren.forPath(s"/$jobId")
    val src = new String(client.getData.forPath(s"/$jobId/repo"))
    val branch = new String(client.getData.forPath(s"/$jobId/branch"))

    val maki: String = loadMaki(src, branch)

    val jobManager: JobManager = createJobManager(maki, jobId, client, attempts, actionsQueue)
    restoreJobState(jobManager, jobId, client)

    jobManager.start()
    jobManager
  }

  def restoreJobState(jobManager: JobManager, jobId: String, client: CuratorFramework): Unit = {

    val tasks = client.getChildren.forPath(s"/$jobId").asScala.toSeq.filter(n => n.startsWith("task"))
    for (task <- tasks) {

      if (client.getData.forPath(s"/$jobId/$task").sameElements(ActionStatus.queued.toString.getBytes) ||
        client.getData.forPath(s"/$jobId/$task").sameElements(ActionStatus.started.toString.getBytes)) {

        jobManager.reQueueAction(task.substring(task.indexOf("task-") + 5))

      }

    }

  }

}

File: leader-common/src/main/kotlin/org/apache/amaterasu/leader/common/execution/JobLoader.kt
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.apache.amaterasu.leader.common.execution

import org.apache.amaterasu.common.configuration.enums.ActionStatus
import org.apache.amaterasu.common.dataobjects.ActionData
import org.apache.amaterasu.common.logging.KLogging
import org.apache.amaterasu.leader.common.dsl.GitUtil
import org.apache.amaterasu.leader.common.dsl.JobParser
import org.apache.curator.framework.CuratorFramework
import org.apache.zookeeper.CreateMode
import java.util.concurrent.BlockingQueue

object JobLoader : KLogging() {

    @JvmStatic
    fun loadJob(src: String, branch: String, jobId: String, userName: String, password: String, client: CuratorFramework, attempts: Int, actionsQueue: BlockingQueue<ActionData>): JobManager {

        // creating the jobs znode and storing the source repo and branch
        client.create().withMode(CreateMode.PERSISTENT).forPath("/$jobId")
        client.create().withMode(CreateMode.PERSISTENT).forPath("/$jobId/repo", src.toByteArray())
        client.create().withMode(CreateMode.PERSISTENT).forPath("/$jobId/branch", branch.toByteArray())

        val maki: String = loadMaki(src, branch, userName, password)

        return createJobManager(maki, jobId, client, attempts, actionsQueue)

    }

    fun createJobManager(maki: String, jobId: String, client: CuratorFramework, attempts: Int, actionsQueue: BlockingQueue<ActionData>): JobManager {

        return JobParser.parse(
                jobId,
                maki,
                actionsQueue,
                client,
                attempts
        )
    }

    fun loadMaki(src: String, branch: String, userName: String = "", password: String = ""): String {

        // cloning the git repo
        log.debug("getting repo: $src, for branch $branch")
        GitUtil.cloneRepo(src, branch, userName, password)

        // parsing the maki.yaml and creating a JobManager to
        // coordinate the workflow based on the file
        val maki = JobParser.loadMakiFile()
        return maki
    }

    @JvmStatic
    fun reloadJob(jobId: String, userName: String, password: String, client: CuratorFramework, attempts: Int, actionsQueue: BlockingQueue<ActionData>): JobManager {

        //val jobState = client.getChildren.forPath(s"/$jobId")
        val src = String(client.data.forPath("/$jobId/repo"))
        val branch = String(client.data.forPath("/$jobId/branch"))

        val maki: String = loadMaki(src, branch, userName, password)

        val jobManager: JobManager = createJobManager(maki, jobId, client, attempts, actionsQueue)
        restoreJobState(jobManager, jobId, client)

        jobManager.start()
        return jobManager
    }

    fun restoreJobState(jobManager: JobManager, jobId: String, client: CuratorFramework): Unit {

        val tasks = client.children.forPath("/$jobId").filter { it.startsWith("task") }

        for (task in tasks) {

            val status = ActionStatus.valueOf(String(client.data.forPath("/$jobId/$task")))
            if (status == ActionStatus.Queued || status == ActionStatus.Started) {
                jobManager.reQueueAction(task.substring(task.indexOf("task-") + 5))
            }

        }

    }


}